### PR TITLE
linkcheck builder: begin using session-based HTTP requests

### LINKED ISSUE #11324: linkcheck builder: begin using requests.Session functionality during linkchecking
**Is your feature request related to a problem? Please describe.**
At the moment, the `linkcheck` builder performs individual `request.get` (or similar HTTP request method) operations during linkchecking, without any explicit connection or session pooling.

This may be inefficient, because it seems likely that for many use cases, linkchecking will make multiple requests to the same host (because documentation references are likely to have host-locality).

**Describe the solution you'd like**
Confirmation that connection pooling is not currently in use would be a good starting point; in other words: we should confirm that linkchecking of multiple URLs on a single host results in multiple TCP connections.  Ideally this should be written as a test case.

If we can confirm that the problem exists, then we may be able to use some of the [`Session` object functionality](https://requests.readthedocs.io/en/latest/user/advanced/#session-objects) from the `requests` library that's already in use here to enable connection pooling.

**Describe alternatives you've considered**
None so far, although open to suggestions (and improvements on the definition of this feature request).

**Additional context**
- See note / suggestion about use of context managers: https://github.com/sphinx-doc/sphinx/issues/11317#issuecomment-1508319197

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix
- Refactoring

### Purpose
- Reduce the number of HTTP connections used during linkchecking, reducing network resource usage

### Detail
- ~~Intersphinx requests and image downloading are also migrated to session-based requests~~
- Resolves #11324.

### Relates
- Builds upon code from #11318.

### CODE REVIEW

**Comment by jayaddison** [comment]
Although it might seem strange, I do expect (and intend) the unit tests to fail with 4305d11facf0c0c088e4b27794f524344e6221f2 to begin with.  Despite the pull request title, the session support (ad12d2578cb14b97f7dadb549e63a2c0a45ab61b) is reverted as of that commit, and so the assertion that [only a single HTTP connection is used](https://github.com/sphinx-doc/sphinx/blob/4305d11facf0c0c088e4b27794f524344e6221f2/tests/test_build_linkcheck.py#L102) should fail.

Reverting 4305d11facf0c0c088e4b27794f524344e6221f2 should re-introduce session support, and allow those tests to pass.

**Comment by jayaddison** [comment]
Argh, I've made a mess of this.  It looks like I forgot to update one of the Intersphinx tests to reflect the other changes.

**Comment by jayaddison** [comment]
Todo: update a leftover reference to `requests.get`: https://github.com/sphinx-doc/sphinx/blob/72e80b76de13fa9d80286cf3c7d745f8e143f2f9/sphinx/transforms/post_transforms/images.py#L75

**Comment by jayaddison** [comment]
> * [x]  Todo: update a leftover reference to `requests.get`: https://github.com/sphinx-doc/sphinx/blob/72e80b76de13fa9d80286cf3c7d745f8e143f2f9/sphinx/transforms/post_transforms/images.py#L75
> 
> 
> Done, although naively; adding some test coverage could be a good idea.

This is better handled during a follow-up cleanup change, I think.

**Review by AA-Turner — APPROVED** [review]

**Comment by AA-Turner** [comment]
Intermittant failures on Python 3.10 with DU 0.18 or 0.19--other than that looks good.

A

File: sphinx/builders/linkcheck.py [hunk 1]
```diff
@@ -424,6 +425,7 @@ def check(docname: str) -> tuple[str, str, int]:
             check_request = self.wqueue.get()
             next_check, hyperlink = check_request
             if hyperlink is None:
+                self._session.close()
```
Comment (AA-Turner) [inline sphinx/builders/linkcheck.py:429]: Why do we close here?

**Comment by jayaddison** [comment]
> Intermittant failures on Python 3.10 with DU 0.18 or 0.19--other than that looks good.

Thanks for the review; I'm taking a look into those test failures now.

FWIW: I saw similar behaviour using Python 3.11 before adding acec7108a788d8ed2f80b93482afb563316c8a25 - and I think that it's possible that calling `flush_headers` during the response may help (but I'll try to confirm that).

File: sphinx/builders/linkcheck.py (hunk 1 above)
Comment (jayaddison) [reply sphinx/builders/linkcheck.py:429]: I wanted to ensure that the `requests` session's resources are cleaned up when the linkchecker completes -- some warnings appeared during testing to indicate that they weren't -- and this seemed to be a good place for it (the `None` hyperlink is used as a marker (sentinel?) to indicate that the worker should shutdown).

**Comment by jayaddison** [comment]
> FWIW: I saw similar behaviour using Python 3.11 before adding https://github.com/sphinx-doc/sphinx/commit/acec7108a788d8ed2f80b93482afb563316c8a25 - and I think that it's possible that calling `flush_headers` during the response may help (but I'll try to confirm that).

On second thoughts: the asserted header is an auth header from the (linkcheck) _client_ - so that makes the server headers seem less-relevant.  Either way: I'll investigate.

**Comment by jayaddison** [comment]
> Intermittant failures on Python 3.10 with DU 0.18 or 0.19--other than that looks good.

There seem to be two problems that contribute to the failures:

- [Reading from the client headers after beginning the handler's communication back to the client](https://github.com/sphinx-doc/sphinx/blob/31ca85ba76b707d022dfae146adc2474453fb2ea/tests/test_build_linkcheck.py#L255-L257) seems unreliable (sometimes an empty header-set is received)
- With [`stream=False` (the default)] for body content reads](https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow), the server may block.  That _might_ indicate a `requests` bug, since for an HTTP HEAD request I don't think that body content should be expected.  An alternative would be to send a zero `Content-Length` header, as we do in another [server handler fixture](https://github.com/sphinx-doc/sphinx/blob/31ca85ba76b707d022dfae146adc2474453fb2ea/tests/test_build_linkcheck.py#L361) -- personally I'd prefer to preserve the two different server behaviours because it provides some test coverage diversity.

I've prepared fixups for each of these -- they're testing well so far -- and may also backport them to #11317 for pedantic consistency.

**Comment by jayaddison** [comment]
> I've prepared fixups for each of these -- they're testing well so far -- and may also backport them to https://github.com/sphinx-doc/sphinx/pull/11318 for pedantic consistency.

Maybe it's a better idea _not_ to backport them.  That would help to confirm whether these issues affect both individual (non-session) HTTP requests, or only session-based HTTP requests.

**Comment by jayaddison** [comment]
> With [stream=False (the default) for body content reads](https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow), the server may block during [this HEAD request](https://github.com/sphinx-doc/sphinx/blob/31ca85ba76b707d022dfae146adc2474453fb2ea/sphinx/builders/linkcheck.py#L331-L333). That might indicate a requests bug, since for an HTTP HEAD request I don't think that body content should be expected. An alternative would be to send a zero Content-Length header, as we do in another [server handler fixture](https://github.com/sphinx-doc/sphinx/blob/31ca85ba76b707d022dfae146adc2474453fb2ea/tests/test_build_linkcheck.py#L361) -- personally I'd prefer to preserve the two different server behaviours because it provides some test coverage diversity.

Even after adjusting the code to use `stream=True`, there still seems to be the possibility of a stalled connection (resulting in intermittent [timeouts](https://github.com/sphinx-doc/sphinx/actions/runs/4764669941/jobs/8469538943#step:10:1945)).

I'm still relatively confident that sending the `Content-Length` header from all of the server test handler fixtures would resolve this.. but I'm still trying to understand the problem in more detail to figure out whether there's a deeper bug that can be resolved, or whether it's unsolvable.

**Comment by jayaddison** [comment]
Taking a break; I'll look at this again later today.

**Comment by jayaddison** [comment]
Todo:

- [ ] Repeat test continuous integration from commit ac2baaa500050ef15f88394751ae7b011e3f715d to see whether the flakiness appears to be resolved by that

**Comment by jayaddison** [comment]
> Todo:
> 
>     * [ ]  Repeat test continuous integration from commit [ac2baaa](https://github.com/sphinx-doc/sphinx/commit/ac2baaa500050ef15f88394751ae7b011e3f715d) to see whether the flakiness appears to be resolved by that

Back to the drawing board(s).  Some tests are still failing (see jayaddison/sphinx#2), including cases that 4d485aee6f304d432121f57c375a479f9674cc5b, 599f1054d38e3760de3b5da5921de596702fb61f and ac2baaa500050ef15f88394751ae7b011e3f715d were intended to resolve.

**Comment by jayaddison** [comment]
> Intermittant failures on Python 3.10 with DU 0.18 or 0.19--other than that looks good.

@AA-Turner @francoisfreitag It seems that #11323 resolves much of the test flakiness here, but I'm (still) reluctant to merge that without understanding why it improves the results so much.

Many of the failures appear as timeouts, and #11323 does skip one client-server roundtrip (HTTP 405 unsupported on `HEAD` request) for each of two links checked during the test.. but I'm not convinced that that should make a big difference.

**Comment by jayaddison** [comment]
> Many of the failures appear as timeouts, and #11323 does skip one client-server roundtrip (HTTP 405 unsupported on `HEAD` request) for each of two links checked during the test.. but I'm not convinced that that should make a big difference.

Could it be that when each [`head(...)` request call](https://github.com/sphinx-doc/sphinx/blob/4a9519757d3480d3bbd8f5d4d59830e5c63aa05b/sphinx/builders/linkcheck.py#L331) fails due to the HTTP 405 unsupported response (because `support_head=False`), then the resulting [`err.response` reference](https://github.com/sphinx-doc/sphinx/blob/4a9519757d3480d3bbd8f5d4d59830e5c63aa05b/sphinx/builders/linkcheck.py#L338) remains in scope, preventing the original request's connection from being returned to the pool and consuming one of a limited number of connections available to the workers, which are eventually exceeded leading to a timeout?

Not entirely convinced by that theory myself, but I suppose it'd be possible to test.

jayaddison/sphinx#2 is currently indicating that the changes here (as of 4a9519757d3480d3bbd8f5d4d59830e5c63aa05b) are stable and do not appear to have intermittent test failures.

**Comment by jayaddison** [comment]
@AA-Turner @francoisfreitag It seems that https://github.com/sphinx-doc/sphinx/pull/11323 resolves much of the test flakiness here, but I'm (still) reluctant to merge that without understanding why it improves the results so much.

Ok: #11349 seems to resolve the same buggy behaviour without requiring a server-side change, so I've merged that here and removed #11323 (in other words: replacing an adjustment to the unit tests with a (preferable, I think) adjustment in the client code).

**Comment by jayaddison** [comment]
Note: I think that this, along with #11349 and #11350, should wait until after the 6.2.0 release for merge.  I'd like for them to bake-in with continuous integration for a while before release, and also I've been churning on them a bit up until today, so that's an indication that they could not be quite ready yet.

(sometimes I feel like I might be stating the obvious, but I'd prefer to do that than cause unnecessary problems)

**Comment by jayaddison** [comment]
I'm planning to rebase this against the master branch, partly as an exercise in reminding myself what the series of changes applied actually is, and what remains relevant to investigate and consider.

**Comment by jayaddison** [comment]
I seem to be running in circles on this one without really figuring out why the remaining unit test failures (all timeouts, I believe) are occuring.

Could it be that multiple unit tests share the same `urllib3` connection pool, and that the workers that each `linkcheck` test spins up aren't being released? (resource contention across tests)

I'll try to test that theory in my personal fork of the repository.

**Comment by jayaddison** [comment]
> I seem to be running in circles on this one without really figuring out why the remaining unit test failures (all timeouts, I believe) are occuring.

:facepalm: I should read before commenting!

The failures don't exhibit timeouts here, but they _do_ exhibit the `test_auth_header_uses_first_match` header-recording problem.

**Comment by jayaddison** [comment]
There was a failure case that I hadn't seen before in https://github.com/sphinx-doc/sphinx/actions/runs/4820054596/jobs/8584030994 from commit 0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5 here.

From the test failure output of [`test_too_many_requests_retry_after_HTTP_date`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L635-L656):

```
E         + Traceback (most recent call last):
E         +   File "C:\hostedtoolcache\windows\Python\3.11.3\x64\Lib\socketserver.py", line 691, in process_request_thread
E         +     self.finish_request(request, client_address)
E         +   File "C:\hostedtoolcache\windows\Python\3.11.3\x64\Lib\socketserver.py", line 361, in finish_request
E         +     self.RequestHandlerClass(request, client_address, self)
E         +   File "C:\hostedtoolcache\windows\Python\3.11.3\x64\Lib\socketserver.py", line 755, in __init__
E         +     self.handle()
E         +   File "C:\hostedtoolcache\windows\Python\3.11.3\x64\Lib\http\server.py", line 434, in handle
E         +     self.handle_one_request()
E         +   File "C:\hostedtoolcache\windows\Python\3.11.3\x64\Lib\http\server.py", line 400, in handle_one_request
E         +     self.raw_requestline = self.rfile.readline(65537)
E         +                            ^^^^^^^^^^^^^^^^^^^^^^^^^^
E         +   File "C:\hostedtoolcache\windows\Python\3.11.3\x64\Lib\socket.py", line 706, in readinto
E         +     return self._sock.recv_into(b)
E         +            ^^^^^^^^^^^^^^^^^^^^^^^
E         + ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
```

I might rebase in a moment, because I think we can drop #11349 from the changeset here.  But I'd like to revisit this failure even if it doesn't reappear immediately.

**Comment by jayaddison** [comment]
I would also like to drop the change from 01c250e7d9e365ff99f4c1222675e388a42d0e59 (reducing the linkcheck worker count for `test_TooManyRedirects_on_HEAD`) if possible, because it seemed like a workaround (again for a client timeout).

**Comment by jayaddison** [comment]
Perhaps a better approach than attempting to merge a set of inter-related changes in a single branch here would be to gradually introduce each one as a separate, simpler pull request.

That would probably help to narrow down the cause of any problems.

Apologies for the noise @AA-Turner @francoisfreitag, and thanks for the review and support.  I'm going to close this and will re-open a few smaller, isolated changes (like adding connection-measurement support in the tests) soon.

The goal continues to be to enable session-based linkchecking.

**Comment by jayaddison** [comment]
> There was a failure case that I hadn't seen before in https://github.com/sphinx-doc/sphinx/actions/runs/4820054596/jobs/8584030994 from commit https://github.com/sphinx-doc/sphinx/commit/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5 here.

I now have local repro of this (`ConnectionResetError`) error, so hopefully that will help to narrow-in on a fix.

**Comment by jayaddison** [comment]
> > There was a failure case that I hadn't seen before in https://github.com/sphinx-doc/sphinx/actions/runs/4820054596/jobs/8584030994 from commit [0ed465c](https://github.com/sphinx-doc/sphinx/commit/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5) here.
> 
> I now have local repro of this (`ConnectionResetError`) error, so hopefully that will help to narrow-in on a fix.

This is an intermittent test failure in [`test_too_many_requests_retry_after_int_delay`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L609-L632) and seems to stop occurring when I remove the [preceding test, `test_TooManyRedirects_on_HEAD`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L570-L588), suggesting that there may be an unintended dependency between those two unit tests.

**Comment by jayaddison** [comment]
> This is an intermittent test failure in [`test_too_many_requests_retry_after_int_delay`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L609-L632) and seems to stop occurring when I remove the [preceding test, `test_TooManyRedirects_on_HEAD`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L570-L588), suggesting that there may be an unintended dependency between those two unit tests.

I think that this indicates that the failure occurs within `test_TooManyRedirects_on_HEAD`; the subsequent test `test_too_many_requests_retry_after_int_delay` just happens to be the runtime context where the (previous test) server thread's exception is raised and printed to standard-error, resulting in that second test failing.

This is confirmed by removing `test_too_many_requests_retry_after_int_delay`; the failure then appears in `test_too_many_requests_retry_after_HTTP_date` (the next test down) instead.

So what's the problem in `test_TooManyRedirects_on_HEAD`?

Well, the test involves a server that will continually redirect the client.

Meanwhile, our test client will close the connection when it reaches some [threshold number of redirects](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L574).

**Comment by jayaddison** [comment]
> Well, the test involves a server that will continually redirect the client.

...only as long as the client performs HTTP HEAD requests, though.  The HTTP GET handler should not perform a redirect...

**Comment by jayaddison** [comment]
> I think that this indicates that the failure occurs within `test_TooManyRedirects_on_HEAD`; the subsequent test `test_too_many_requests_retry_after_int_delay` just happens to be the runtime context where the (previous test) server thread's exception is raised and printed to standard-error, resulting in that second test failing.

Is that a bug in `pytest`'s `capsys` fixture, I wonder?  (the server should have closed down during `test_TooManyRedirects_on_HEAD`.. so the message should have been printed to standard-error during that.. I think?)

**Comment by jayaddison** [comment]
> > This is an intermittent test failure in [`test_too_many_requests_retry_after_int_delay`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L609-L632) and seems to stop occurring when I remove the [preceding test, `test_TooManyRedirects_on_HEAD`](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L570-L588), suggesting that there may be an unintended dependency between those two unit tests.
> 
> I think that this indicates that the failure occurs within `test_TooManyRedirects_on_HEAD`; the subsequent test `test_too_many_requests_retry_after_int_delay` just happens to be the runtime context where the (previous test) server thread's exception is raised and printed to standard-error, resulting in that second test failing.
> 
> This is confirmed by removing `test_too_many_requests_retry_after_int_delay`; the failure then appears in `test_too_many_requests_retry_after_HTTP_date` (the next test down) instead.
> 
> So what's the problem in `test_TooManyRedirects_on_HEAD`?
> 
> Well, the test involves a server that will continually redirect the client.
> 
> Meanwhile, our test client will close the connection when it reaches some [threshold number of redirects](https://github.com/sphinx-doc/sphinx/blob/0ed465c112e4e10f775c7ddfbede1cf1a7ffb9a5/tests/test_build_linkcheck.py#L574).

Ok, I think I've found the problem.  It is straightforward to replicate a `ConnectionResetError` from `test_TooManyRedirects_on_HEAD` when:

- Threaded HTTP servers are in use within the tests
- The test HTTP servers speak the HTTP/1.1 protocol
- Session-based requests are not in use
- Context-managed requests are not in use

The problem occurs after:

- The test client follows multiple HTTP HEAD redirects (that's fine)
- The test client exceeds the max-redirect limit of 5 (also fine)
- The test client issues an HTTP GET fallback in streaming mode (also OK)
- The test server responds HTTP 200 OK (still fine up until now)
- The test client disconnects, because it was in streaming mode and linkchecking is complete
- The test server [writes a response body](https://github.com/jayaddison/sphinx/blob/bd2ed0bfd5a89d3dd2efaad8dd6cbe1519f9037d/tests/test_build_linkcheck.py#L567) that will never be read

The last two items may occur in the inverse order and the result will be the same.

The `Content-Length` header in this case does not help; the client, by design, does not intend to read the remainder of the server response.

I think it'd be possible to update the [`InfiniteRedirectOnHeadHandler` test server handler](https://github.com/jayaddison/sphinx/blob/bd2ed0bfd5a89d3dd2efaad8dd6cbe1519f9037d/tests/test_build_linkcheck.py#L554) to [close the server-side connection after the response is written](https://github.com/jayaddison/sphinx/blob/bd2ed0bfd5a89d3dd2efaad8dd6cbe1519f9037d/tests/test_build_linkcheck.py#L749) to fix this, without changing the test's semantics.

**Comment by jayaddison** [comment]
> Ok, I think I've found the problem. It is straightforward to replicate a ConnectionResetError from test_TooManyRedirects_on_HEAD when:
> ...

Here's an attempt to create a minimal repro case for this problem:

```python
from contextlib import contextmanager
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import time
from threading import Thread

import pytest


class TestHandler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def do_GET(self):
        content = b"content"
        self.send_response(200, "OK")
        self.send_header("Content-Length", len(content))
        self.end_headers()
        self.wfile.write(content)
        self.wfile.flush()


class TestServer(Thread):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.server = ThreadingHTTPServer(("localhost", 7777), TestHandler)

    def run(self):
        self.server.serve_forever()

    def join(self):
        self.server.shutdown()
        self.server.server_close()
        super().join()


@contextmanager
def server_context():
    server_thread = TestServer()
    server_thread.start()
    yield server_thread
    server_thread.join()


@pytest.mark.sphinx('linkcheck', testroot='linkcheck-localserver', freshenv=True)
def test_a(app, capsys):
    with server_context():
        app.build()


@pytest.mark.sphinx('linkcheck', testroot='linkcheck-localserver', freshenv=True)
def test_b(app, capsys):
    time.sleep(0.5)

    _stdout, stderr = capsys.readouterr()
    assert stderr == ""
```

A few notes:

- The problem only seems to occur when the same app configured is re-used between the two unit tests, as specified using `pytest.mark.sphinx` decoration
- The second unit test (where the error appears in the stderr output, but not where the stacktrace originates) does not have to run the linkchecker - it is enough for it to `time.sleep` for a long-enough duration

I'm going to try to simplify the conditions further, likely by temporarily removing code from the `pytest` marker and linkchecker code.

**Comment by jayaddison** [comment]
Ok, after further reduction of the repro case, I don't think that anything about the `ConnectionResetError` / inter-unit-test-log-capture issue is specific to Sphinx.  Here is a repro case where I can reliably reproduce the problem using cPython 3.11.2:

```
from contextlib import contextmanager
import gc
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from threading import Thread
import time

import pytest
import requests


class TestHandler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def do_GET(self):
        content = b"content"
        self.send_response(200, "OK")
        self.send_header("Content-Length", len(content))
        self.end_headers()
        self.wfile.write(content)
        self.wfile.flush()


class TestServer(Thread):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.server = ThreadingHTTPServer(("localhost", 8080), TestHandler)

    def run(self):
        self.server.serve_forever()

    def join(self):
        self.server.shutdown()
        self.server.server_close()
        super().join()


@contextmanager
def server_context():
    server_thread = TestServer()
    server_thread.start()
    yield server_thread
    server_thread.join()


def test_a():
    with server_context():
        requests.get('http://localhost:8080', stream=True)


def test_b(capsys):
    gc.collect()
    time.sleep(0.1)

    _stdout, stderr = capsys.readouterr()
    assert stderr == ""
```

Further notes:

- Introducing `gc` to request garbage collection makes the issue more repeatable using cPython.
- Therefore the theory would be: the `recv_into` exception relates to socket-handling (a thread?) that is deferred until the socket resources are cleaned up; that can occur outside of the unit test where those resources were created.
- The behaviour here may be OS-specific, since some `urllib3` and socket implementation functionality differs by platform.
- There may not be much that `pytest` could do to handle this, since garbage-collecting during teardown of every test would likely introduce significant overhead.

I'm not sure that there's anything else to do for this particular behaviour, given that we can fix it in the unit tests by having the server set `self.close_connection = True` during request handling (in `do_GET`).

It's potentially a slightly-annoying cause of noise in server logs, but everything is behaving to-spec, as far as I can tell (maybe that could indicate an HTTP spec limitation, or under-definition of the spec somewhere - I'm not completely sure).

**Pull request closed without merging**